modules/cdc-ext/src/main/java/org/apache/ignite/cdc/conflictresolve/CacheVersionConflictResolverImpl.java

```diff
@@ -0,0 +1,160 @@
+    }
+
+    /** {@inheritDoc} */
+    @Override public <K, V> GridCacheVersionConflictContext<K, V> resolve(
+        CacheObjectValueContext ctx,
+        GridCacheVersionedEntryEx<K, V> oldEntry,
+        GridCacheVersionedEntryEx<K, V> newEntry,
+        boolean atomicVerComparator
+    ) {
+        GridCacheVersionConflictContext<K, V> res = new GridCacheVersionConflictContext<>(ctx, oldEntry, newEntry);
+
+        if (isUseNew(ctx, oldEntry, newEntry))
+            res.useNew();
+        else {
+            log.warning("Skip update due to the conflict [key=" + newEntry.key() + ", fromCluster=" + newEntry.dataCenterId()
+                + ", toCluster=" + oldEntry.dataCenterId() + ']');
+
+            res.useOld();
+        }
+
+        return res;
+    }
+
+    /**
+     * @param ctx Context.
+     * @param oldEntry Old entry.
+     * @param newEntry New entry.
+     * @param <K> Key type.
+     * @param <V> Key type.
+     * @return {@code True} is should use new entry.
+     */
+    @SuppressWarnings({"unchecked", "rawtypes"})
+    private <K, V> boolean isUseNew(
+        CacheObjectValueContext ctx,
+        GridCacheVersionedEntryEx<K, V> oldEntry,
+        GridCacheVersionedEntryEx<K, V> newEntry
+    ) {
+        if (newEntry.dataCenterId() == clusterId) // Update made on the local cluster always win.
+            return true;
+
+        if (oldEntry.isStartVersion()) // New entry.
+            return true;
+
+        if (oldEntry.dataCenterId() == newEntry.dataCenterId())
+            return newEntry.version().compareTo(oldEntry.version()) > 0; // New version from the same cluster.
+
+        if (conflictResolveFieldEnabled) {
+            Object oldVal = oldEntry.value(ctx);
+            Object newVal = newEntry.value(ctx);
+
+            if (oldVal != null && newVal != null) {
+                Comparable o;
+                Comparable n;
+
+                try {
+                    if (oldVal instanceof BinaryObject) {
+                        o = ((BinaryObject)oldVal).field(conflictResolveField);
+                        n = ((BinaryObject)newVal).field(conflictResolveField);
+                    }
+                    else {
+                        o = U.field(oldVal, conflictResolveField);
+                        n = U.field(newVal, conflictResolveField);
+                    }
+
+                    if (o == null || n == null)
+                        return o == null;
+
+                    return o.compareTo(n) < 0;
+                }
+                catch (Exception e) {
+                    log.error("Error while resolving replication conflict with field '" +
+                        conflictResolveField + "'.\nConflict resolve with the field disabled.", e);
+
+                    conflictResolveFieldEnabled = false;
+                }
+            }
+        }
+
+        // Cluster with the lower ID have biggest priority(e.g. first cluster is main).
+        return newEntry.dataCenterId() < oldEntry.dataCenterId();
```

Comment:
> Warn message with the key should be also generated here since the decision made by fuzzy logic.

Conflict resolver is extremely hot code so I don't think we can insert logging here.

> BTW, do we able to mention keys in logs in theory?

Yes, please, take a look at `IgniteSystemProperties#IGNITE_TO_STRING_INCLUDE_SENSITIVE`

Reply:
> Conflict resolver is extremely hot code so I don't think we can insert logging here.

This is the rarest case when hot cases skipped.

> Yes, please, take a look at IgniteSystemProperties#IGNITE_TO_STRING_INCLUDE_SENSITIVE

This flag works only when you use it %) 
For example you forget to use it at CacheVersionConflictResolverImpl#resolve

Reply:
Value of this flag analyzed during `toString` invocation.
Please, take a look at `CacheObjectAdapter#toString` and delaration of the `val` field:
```
public abstract class CacheObjectAdapter ... {
    @GridToStringInclude(sensitive = true) //Note sensitive is true.
    @GridDirectTransient
    protected Object val;

   ...
}
```

Reply:
> Value of this flag analyzed during toString invocation.

Great!

So, only warn required here.